src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,972 @@
+            var newMethodDeclaration = codeGenerationService.CreateMethodDeclaration(newMethod, options: new CodeGenerationOptions(options: options, parseOptions: expression.SyntaxTree.Options));
+            return newMethodDeclaration;
+        }
+
+        /// <summary>
+        /// This method goes through all the invocation sites and adds a new argument with the expression to be added.
+        /// It also introduces a parameter at the original method site.
+        /// 
+        /// Example:
+        /// public void M(int x, int y)
+        /// {
+        ///     int f = [|x * y|];
+        /// }
+        /// 
+        /// public void InvokeMethod()
+        /// {
+        ///     M(5, 6);
+        /// }
+        /// 
+        /// ---------------------------------------------------->
+        /// 
+        /// public void M(int x, int y, int f) // parameter gets introduced
+        /// {
+        /// }
+        /// 
+        /// public void InvokeMethod()
+        /// {
+        ///     M(5, 6, 5 * 6); // argument gets added to callsite
+        /// }
+        /// </summary>
+        private async Task<SyntaxNode> ModifyDocumentInvocationsAndIntroduceParameterAsync(Compilation compilation,
+            Document originalDocument, Document document, Dictionary<TIdentifierNameSyntax, IParameterSymbol> mappingDictionary,
+            SyntaxNode containingMethod, TExpressionSyntax expression,
+            bool allOccurrences, string parameterName, int insertionIndex,
+            List<SyntaxNode> invocations, CancellationToken cancellationToken)
+        {
+            var semanticFacts = document.GetRequiredLanguageService<ISemanticFactsService>();
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var editor = new SyntaxEditor(root, generator);
+            var invocationSemanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            foreach (var invocation in invocations)
+            {
+                var expressionEditor = new SyntaxEditor(expression, generator);
+
+                var invocationArguments = invocation is TObjectCreationExpressionSyntax
+                    ? syntaxFacts.GetArgumentsOfObjectCreationExpression(invocation)
+                    : syntaxFacts.GetArgumentsOfInvocationExpression(invocation);
+
+                var parameterToArgumentMap = MapParameterToArgumentsAtInvocation(semanticFacts, invocationArguments, invocationSemanticModel, cancellationToken);
+
+                var argumentListSyntax = invocation is TObjectCreationExpressionSyntax
+                    ? syntaxFacts.GetArgumentListOfObjectCreationExpression(invocation)
+                    : syntaxFacts.GetArgumentListOfInvocationExpression(invocation);
```

Comment:
move this above, then use the helper to get the arguments of an argumentlist.

Reply:
fixed